azext_iot/central/providers/device_provider.py

```diff
@@ -194,20 +201,92 @@ def get_device_registration_info(
+    def filter_dps_info(self, dps_info, device_status):
+        error = {
+            "provisioned": "None",
+            "registered": "Device it not yet provisioned.",
+            "blocked": "Device is blocked by admin",
+            "unassociated": "Device does not have a valid template associated with it",
+        }
+
+        filtered_dps_info = {
+            "status": dps_info.get("status"),
+        }
+
+        filtered_dps_info.update({"error": error.get(device_status)})
+        return filtered_dps_info
+
+    def update_device_info_filter_status(self, device, value):
+        if value is None:
+            return self.update_device_status(device)
+        updated_device_data = {
+            "id": device["id"],
+            "displayName": device.get("displayName"),
+            "instanceOf": device.get("instanceOf"),
+            "simulated": device.get("simulated"),
+            "deviceStatus": value,
+        }
+        return updated_device_data
+
+    def update_device_status(self, device):
+        if device["approved"] is False:
+            updated_device = self.update_device_info_filter_status(
+                device, DeviceStatus.blocked.value
+            )
+        else:
+            if device.get("instanceOf") is None:
+                updated_device = self.update_device_info_filter_status(
+                    device, DeviceStatus.unassociated.value
+                )
+            else:
+                if device["provisioned"] is False:
+                    updated_device = self.update_device_info_filter_status(
+                        device, DeviceStatus.registered.value
+                    )
+                else:
+                    updated_device = self.update_device_info_filter_status(
+                        device, DeviceStatus.provisioned.value
+                    )
+        return updated_device
+
+    def update_all_device_status(self, devices):
+        filtered_device_list = []
+        for device in devices:
+            filtered_device_list.append(self.update_device_status(device))
+        return filtered_device_list
+
+    def filter_device_list(self, devices, device_status):
+        filtered_device_list = []
+        updated_device_list = self.update_all_device_status(devices)
+        if device_status is None:
+            return updated_device_list
+
+        for device in updated_device_list:
+            if device.get("deviceStatus") is device_status:
+                filtered_device_list.append(device)
+        return filtered_device_list
+
+    def get_all_registration_info(
+        self, device_status, central_dns_suffix="azureiotcentral.com"
+    ):
+
         logger.warning("This command may take a long time to complete execution.")
         devices = self.list_devices(central_dns_suffix=central_dns_suffix)
         real_devices = [
             device for device in devices.values() if not device["simulated"]
         ]
-        if len(devices) != len(real_devices):
+
```

Comment:
I think the flow here should be something like:

```
real_devices_with_status = [self.add_device_status(device) for device in real_devices]

filtered_devices = real_devices_with_status
if device_status:
  filtered_devices = [
      device 
      for device in devices
      # remove devices without central device provisioning information
      # as that means the device does not have DPS information
      if device.get("deviceStatus") and device.get("deviceStatus") == device_status
  ]
```

Reply:
Restricted the call to devices that are provisioned.